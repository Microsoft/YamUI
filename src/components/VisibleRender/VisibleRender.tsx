/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import Observer from 'react-intersection-observer';
import { BaseComponentProps } from '../../util/BaseComponent/props';

export interface VisibleRenderProps extends BaseComponentProps {
  /**
   * A callback which will be triggered when the component is scrolled into view.
   */
  onEnter?: () => void;

  /**
   * A callback which will be triggered when the component is scrolled out of view.
   */
  onLeave?: () => void;

  /**
   * Render prop to return child content when the component is visible.
   */
  whenInView?: () => React.ReactNode;

  /**
   * Render prop to return child content when the component is visible.
   */
  whenOutOfView?: () => React.ReactNode;

  /**
   * Wrapper element tag name.
   * @default 'div'
   */
  tag?: string;

  /**
   * A CSS margin string which pushes the intersection boundary further in or out of the viewport.
   * A positive value will expand the viewport threshold so your component is considered "in view" while it's
   * still offscreen, for example you could set thresholdOffset="200px" you want an image to start loading before
   * it scrolls into view. A negative value will move the boundary into the viewport, triggering "in view" after
   * it's already becoming visible.
   * "200px", "200px 0 50px 0", "-50px" are all valid values.
   */
  rootMargin?: string;
}

/**
 * VisibleRender allows for conditional child rendering and callbacks when scrolled in or out of view.
 */
export default class VisibleRenderer extends React.Component<VisibleRenderProps> {
  public render() {
    const { rootMargin, tag } = this.props;

    return (
      <Observer
        tag={tag || 'div'}
        rootMargin={rootMargin}
        onChange={this.onVisibilityChange}
        render={this.getObserverChildren}
      />
    );
  }

  private getObserverChildren = (isVisible: boolean) => {
    const { whenInView, whenOutOfView } = this.props;
    if (isVisible && whenInView) {
      return whenInView();
    }

    if (!isVisible && whenOutOfView) {
      return whenOutOfView();
    }

    return null;
  };

  private onVisibilityChange = (isVisible: boolean) => {
    if (isVisible) {
      this.props.onEnter && this.props.onEnter();
    } else {
      this.props.onLeave && this.props.onLeave();
    }
  };
}
