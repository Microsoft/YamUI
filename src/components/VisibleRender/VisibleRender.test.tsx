/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import Observer from 'react-intersection-observer';
import { shallow, ShallowWrapper } from 'enzyme';
import VisibleRender, { VisibleRenderProps } from './index';

type CallbackProp = ((inView: boolean) => void);
type RenderProp = ((inView: boolean) => React.ReactNode);

describe('<VisibleRender />', () => {
  let component: ShallowWrapper<VisibleRenderProps, {}>;
  const renderInView = () => <span>IN VIEW</span>;
  const renderOutOfView = () => <span>OUT OF VIEW</span>;
  let onEnterCallback: jest.Mock<{}>;
  let onLeaveCallback: jest.Mock<{}>;
  let renderProp: RenderProp;
  let callbackProp: CallbackProp;

  describe('with minimal options', () => {
    beforeEach(() => {
      component = shallow(<VisibleRender />);
      renderProp = component.find(Observer).prop('render') as RenderProp;
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('renders empty content', () => {
      expect(renderProp(true)).toMatchSnapshot();
    });
  });

  describe('with all options', () => {
    beforeEach(() => {
      onEnterCallback = jest.fn();
      onLeaveCallback = jest.fn();

      component = shallow(
        <VisibleRender
          whenInView={renderInView}
          whenOutOfView={renderOutOfView}
          onEnter={onEnterCallback}
          onLeave={onLeaveCallback}
          rootMargin="100px"
          tag="span"
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('Observer rendering', () => {
      let renderProp: RenderProp;

      beforeEach(() => {
        renderProp = component.find(Observer).prop('render') as RenderProp;
      });

      it('in view render matches its snapshot', () => {
        expect(renderProp(true)).toMatchSnapshot();
      });

      it('out of view render matches its snapshot', () => {
        expect(renderProp(false)).toMatchSnapshot();
      });
    });

    describe('Observer callbacks', () => {
      beforeEach(() => {
        callbackProp = component.find(Observer).prop('onChange') as CallbackProp;
      });

      describe('when in view', () => {
        beforeEach(() => {
          callbackProp(true);
        });

        it('in view callback is properly called', () => {
          expect(onEnterCallback).toHaveBeenCalled();
        });
      });

      describe('when out of view', () => {
        beforeEach(() => {
          callbackProp(false);
        });

        it('out of view callback is properly called', () => {
          expect(onLeaveCallback).toHaveBeenCalled();
        });
      });
    });
  });
});
