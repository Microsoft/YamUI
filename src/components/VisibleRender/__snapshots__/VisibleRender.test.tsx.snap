// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisibleRender /> with all options Observer rendering in view render matches its snapshot 1`] = `
<span>
  IN VIEW
</span>
`;

exports[`<VisibleRender /> with all options Observer rendering out of view render matches its snapshot 1`] = `
<span>
  OUT OF VIEW
</span>
`;

exports[`<VisibleRender /> with all options matches its snapshot 1`] = `
<Observer
  onChange={[Function]}
  render={[Function]}
  rootMargin="100px"
  tag="span"
  threshold={0}
  triggerOnce={false}
/>
`;

exports[`<VisibleRender /> with minimal options matches its snapshot 1`] = `
<Observer
  onChange={[Function]}
  render={[Function]}
  tag="div"
  threshold={0}
  triggerOnce={false}
/>
`;

exports[`<VisibleRender /> with minimal options renders empty content 1`] = `null`;
